### DISCUSSION TITLE
Initial delay

### DISCUSSION THREAD

**Discussion by psh05** [spring-projects/spring-kafka#2147] [Q&A]
@tomazfernandes Is it possible to set an initial delay of x hours then set a different delay for subsequent retries?

**Comment by tomazfernandes** [comment]
Currently I don't think we can, but I think we could add a `delayed` property to the builder and annotation and add that to the first back off.

If @garyrussell is ok with this, I think you can open an issue for this enhancement.

Thanks for the suggestions!

**Comment by psh05** [comment]
Just to clarify when I said "then set a different delay for subsequent retries" - I meant e.g. 24 hours for the first delay followed by a retry of 5m, 1 hr, 2 hr.  
Will wait for @garyrussell to confirm then happy to create an issue! 
Thanks!

**Comment by tomazfernandes** [comment]
Ok, so for example `@RetryableTopic(initialDelay="1d", @BackOff(delay="5m", multiplier="2.0"))`, would result in the first delay of 24h, then 5m, 10m, 20m, and so on for the number of attempts. Is that correct? 

Of course, the first delayed attempt also counts as an attempt, so if it's 5 attempts, would be 0, 24h, 5m, 10m, 20m.

**Reply by tomazfernandes** [reply]
The 0 is for the first attempt in the main topic which has no BackOff

**Reply by psh05** [reply]
So if we have application A that puts a message onto topic-1. Application B can only process that after 24 hrs initially then retry after 5m, 1hr, 8hr if a failure occurs. 

To do this, App B would consume from topic-1 and essentially do nothing/throw an exception to mimic a failure if the retry number is 0 (do we know the retry number from the metadata?), then for subsequent retries (i.e. after 24 hours and 5mins etc) process as usual. Is that correct?

With the '5m, 1hr, 8hr' I'm trying to understand if a custom delay like that is possible. Otherwise initial delay followed by a backoff as you mentioned is sufficient (24h, then 5m, 10m, 20m).

**Reply by psh05** [reply]
Of course this is one scenario I'm trying to solve, there are others where what you mentioned with  0, 24h, 5m, 10m, 20m is perfect.

**Reply by tomazfernandes** [reply]
Oh, ok, I see, you want to have custom values. Good news then, you can have that with current logic. 

I think it wouldn't actually be hard to implement having custom values set through the APIs; not sure though if that's something many people would want, since there's no implementation on Spring Retry for example and people seem to be ok with that.

Currently, it can be done if you're using the builder. You'd have to implement a `CustomBackOffPolicy` such as:

```
public static class CustomBackOffPolicy implements SleepingBackOffPolicy<CustomBackOffPolicy> {

	private final List<Long> backOffValues;
	private Sleeper sleeper = new ThreadWaitSleeper();

	public CustomBackOffPolicy(List<Long> backOffValues) {
		this.backOffValues = backOffValues;
	}

	@Override
	public CustomBackOffPolicy withSleeper(Sleeper sleeper) {
		this.sleeper = sleeper;
		return this;
	}

	@Override
	public BackOffContext start(RetryContext context) {
		return new CustomBackOffContext(backOffValues);
	}

	@Override
	public void backOff(BackOffContext backOffContext) throws BackOffInterruptedException {
		CustomBackOffContext customBackOffContext = (CustomBackOffContext) backOffContext;
		try {
			sleeper.sleep(customBackOffContext.getNextBackoff());
		} catch (InterruptedException e) {
			Thread.interrupted();
			throw new BackOffInterruptedException("Interrupting while backing off", e);
		}
	}

	static class CustomBackOffContext implements BackOffContext {
		private final List<Long> customBackOffs;

		int index = 0;

		CustomBackOffContext(List<Long> customBackOffs) {
			this.customBackOffs = customBackOffs;
		}

		Long getNextBackoff() {
			if (index == customBackOffs.size()) {
				throw new IllegalStateException("Attempting more back offs then values provided. " +
						"Number of attempts: " + (index + 1) + " Values provided: " + this.customBackOffs);
			}
			return customBackOffs.get(index++);
		}
	}
}

```

And then:

```
CustomBackOffPolicy customBackOffPolicy = new CustomBackOffPolicy(Arrays.asList(Duration.ofDays(1).toMillis(),
	Duration.ofMinutes(5).toMillis(),
	Duration.ofHours(1).toMillis(), Duration.ofHours(8).toMillis()));

return RetryTopicConfigurationBuilder
	.newInstance()
	.customBackoff(customBackOffPolicy)
	.maxAttempts(5)
        .create(template)
```

Maybe we could consider having this class as part of the framework, since it doesn't pollute the APIs and can be useful.

**Reply by garyrussell** [reply]
Sorry @psh05 I misunderstood your request in the email (as I said there; best to ask such questions in globally accessible areas like this or stack overflow; I don't work on Fridays and I scanned your question quickly).

I interpreted "initial delay" as a delay for the first attempt, not the first retry; sorry.

There should be no problems with long delays; however, if you later reduce them, any retries "behind" a long delayed record in the same partition will still get the remainder of that delay.

Again, we can convert this to an issue.

**Reply by psh05** [reply]
Hi @garyrussell I guess it can also be an initial delay on the first attempt - so rather than doing "nothing" in the first attempt then processing the message on the first retry (which would be after 24 hours), we could have an initial delay of 24 hours i.e. the first attempt is after 24 hours, and following that the retries can follow either a custom policy or a back off etc which might be cleaner. i.e  24h, 5m, 10m, 20m

**Reply by garyrussell** [reply]
That would add some complexity; I would rather not introduce that at this time.

Having said that, @tomazfernandes wouldn't it work if the producer simply set the timestamp on the initial record.

**Reply by tomazfernandes** [reply]
Sure, having access to the producer, since the main topic is already wrapped in a `KafkaBackoffAwareMessageListenerAdapter`, by setting the default `retry_topic-backoff-timestamp` header in the record with the time we want the record to be processed, the consumer will back the partition off until then.

As for the initial attempt delay, I do have some ideas around that, but I also prefer deferring it to a later moment.

@garyrussell, DYT it's worth providing an `ArbitraryBackOffPolicy` or similar such as the implementation above so users can set arbitrary delay values OOTB? We might perhaps add that to the builder. Or maybe it's a niche use case and we can leave the implementations up to the users as is today. I'm ok with both directions.

Thanks

**Reply by psh05** [reply]
Not sure if it work as it's produced by a different application and there are multiple consumers (some which require immediate processing) hence adding the backoff timestamp from the producer wouldn't work (if I've understood correctly)

**Reply by garyrussell** [reply]
I would stick with what we have since @psh05 meant the initial delay for the first retry not for the first delivery.

Let's wait to see if someone actually wants it.

**Reply by tomazfernandes** [reply]
Sure. To be clear, I mean enabling custom values for the retries as @psh05 said - not touching first delivery.

By the way, @psh05, did the custom BackOff solution work for you?

**Reply by garyrussell** [reply]
Sorry - too many balls in the air - I think providing the hook for a custom back off should be enough at this time.

**Reply by tomazfernandes** [reply]
No worries, sounds good, thanks @garyrussell

**Reply by psh05** [reply]
Yeah @tomazfernandes the custom delay worked, thanks for that! (Currently testing it with a longer delay so will let you know if I have any issues) 

1. I think it would be good to have the initial delay with a backoff (as the example you gave - don’t think you can currently do that with a custom backoff) and an initial delay with a fixed delay (e.g. 24 hr, followed by 5x 1 hr. This could be done with the custom delay however you have to define the 1 hour delay 5 times and this is just an example). Are you happy to proceed with an issue for this?

3. I tried to retrieve the delivery attempt using the DELIVERY_ATTEMPT header but that didn’t seem to work (no messages were consumed and no errors). Any ideas or if there is a different way to get the retry/delivery number?

**Reply by psh05** [reply]
@tomazfernandes @garyrussell Sorry to chase! Any thoughts on my last comment?

**Reply by tomazfernandes** [reply]
Hi @psh05, sorry, some other issues came along and I didn't reply.

I'm not sure about having this `initialDelay` setting in the annotation considering the same result can be achieved with a custom backoff policy - I think it's usually better not to pollute the annotation unless really necessary.

If you need to automate the delay values creation, you can use the `BackOffValuesGenerator` class - you provide the `BackOffPolicy` and number of attempts, and it'll return a list of delays. Then you can add your initial delay and pass on to the custom back off.

But I'd like to hear @garyrussell's thoughts.

For the attempts count, the attempts information is in the `RetryTopicHeaders.DEFAULT_HEADER_ATTEMPTS` header (`retry_topic-attempts`).

Let me know if that works for you. 

Thanks

**Comment by Ramenbet-giris** [comment]
Yes, it's possible to set an initial delay of x hours and then use a different delay for subsequent retries. Most job processing frameworks support this through custom delay logic or retry configurations.

**Comment by Ramenbet-giris** [comment]
For example, you can configure the first delay manually (e.g., 2 hours) and apply a retry strategy for later attempts (e.g., exponential backoff or fixed intervals). This is often done with hooks or middleware in frameworks like Sidekiq, BullMQ, or Celery.

**Comment by Ramenbet-giris** [comment]
If your framework doesn’t natively support this, you can programmatically check the attempt count and adjust the delay accordingly. Let me know your specific use case or framework for further details!

**Comment by PlayF0rtuna** [comment]
Yes, you can set an initial delay of x hours and a different delay for subsequent retries in most job processing frameworks by combining manual configuration and retry strategies.

**Comment by PlayF0rtuna** [comment]
For the initial delay, configure it explicitly (e.g., using a countdown or delay setting). For retries, use a backoff strategy or custom retry logic to adjust the delay dynamically.

**Comment by PlayF0rtuna** [comment]
If the framework lacks native support, you can write custom logic to check retry counts and apply specific delays for the first and subsequent attempts. Let me know your setup for more details!
